modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryFutureAdapter.java

```diff
@@ -338,55 +256,83 @@ public void onPage(@Nullable UUID nodeId, @Nullable Collection<?> data, @Nullabl
                                 "query", qry, true),
                             err));
 
-                    onPage(nodeId, true);
-
-                    notifyAll();
+                    lock.notifyAll();
                 }
             else {
                 if (data == null)
                     data = Collections.emptyList();
 
-                data = dedupIfRequired((Collection<Object>)data);
+                data = dedupIfRequired(data);
+
+                if (qry.query().type() == GridCacheQueryType.TEXT) {
+                    ArrayList unwrapped = new ArrayList();
+
+                    for (Object o: data) {
+                        CacheEntryWithPayload e = (CacheEntryWithPayload) o;
+
+                        Object uKey = CacheObjectUtils.unwrapBinary(
+                            cctx.cacheObjectContext(), e.getKey(), qry.query().keepBinary(), true, null);
+
+                        Object uVal = CacheObjectUtils.unwrapBinary(
+                            cctx.cacheObjectContext(), e.getValue(), qry.query().keepBinary(), true, null);
 
-                data = cctx.unwrapBinariesIfNeeded((Collection<Object>)data, qry.query().keepBinary());
+                        if (uKey != e.getKey() || uVal != e.getValue())
+                            unwrapped.add(new CacheEntryWithPayload<>(uKey, uVal, e.payload()));
+                        else
+                            unwrapped.add(o);
+                    }
+
+                    data = unwrapped;
+
+                } else
+                    data = cctx.unwrapBinariesIfNeeded((Collection<Object>)data, qry.query().keepBinary());
 
-                synchronized (this) {
-                    enqueue(data);
+                synchronized (lock) {
```

Comment:
Yes, it requires for syncing `queue`, `rcvd` and `fut.isDone` to avoid duplicate requests sending for nodes that finish query. Also there is required to use shared lock for onError.